backend/src/config.rs

```diff
@@ -1,16 +1,138 @@
     /// Return either very permissive [`CORS`](`CorsLayer`) configuration
     /// or empty one based on `cors_allow_all` field
     fn get_cors_configuration(&self) -> CorsLayer;
 }
 
+pub trait LoggerExt {
+    /// Initialize logger.
+    ///
+    /// Returns [`WorkerGuard`]s for off-thread writers.
+    /// Should not be dropped.
+    ///
+    /// # Errors
+    ///
+    /// Function returns error if `init_file_rotate` fails
+    fn init_logger(&self) -> Result<Vec<WorkerGuard>, LoggerError>;
+
+    /// Returns [`std:io::Write`] object that rotates files on write
+    ///
+    /// # Errors
+    ///
+    /// Function returns error if `log_file` is not specified
+    fn init_file_rotate(&self) -> Result<FileRotate<AppendTimestamp>, LoggerError>;
+
+    /// Returns non-blocking file writer
+    ///
+    /// Also returns [`WorkerGuard`] for off-thread writing.
+    /// Should not be dropped.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the file logger configuration is empty, file logging
+    /// is disabled or logs filename is not specified
+    fn non_blocking_file_writer(&self) -> Result<(NonBlocking, WorkerGuard), LoggerError>;
+
+    /// Returns non-blocking stdout writer
+    ///
+    /// Also returns [`WorkerGuard`] for off-thread writing.
+    /// Should not be dropped.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the stdout logger configuration is empty or stdout logging
+    /// is disabled
+    fn non_blocking_stdout_writer(&self) -> Result<(NonBlocking, WorkerGuard), LoggerError>;
+}
+
 impl ConfigExt for Config {
     fn get_cors_configuration(&self) -> CorsLayer {
         self.cors_allow_all
             .then_some(CorsLayer::very_permissive())
             .unwrap_or_default()
     }
 }
+
+impl LoggerExt for LoggerConfig {
+    fn init_logger(&self) -> Result<Vec<WorkerGuard>, LoggerError> {
+        let mut guards = Vec::with_capacity(2);
+
+        let mut layers_iter = [
+            self.non_blocking_file_writer().ok(),
+            self.non_blocking_stdout_writer().ok(),
+        ]
+        .into_iter()
+        .flatten()
+        .map(|(writer, guard)| {
+            guards.push(guard);
+            tracing_subscriber::fmt::layer()
+                .with_writer(writer)
+                .with_filter(LevelFilter::from_level(self.trace_level))
+        });
+
+        if let Some(first_layer) = layers_iter.next() {
+            tracing_subscriber::registry()
+                .with(layers_iter.fold(first_layer.boxed(), |layer, next_layer| {
+                    layer.and_then(next_layer).boxed()
+                }))
+                .init();
+        };
+
+        Ok(guards)
+    }
+
+    fn init_file_rotate(&self) -> Result<FileRotate<AppendTimestamp>, LoggerError> {
+        let config = self.file.as_ref().ok_or(LoggerError::EmptyConfig)?;
+        Ok(FileRotate::new(
+            config.log_file.as_ref().ok_or(LoggerError::NoFileName)?,
```

Comment:
I suggest to check additionally, that the path is not empty and return the same error if it is

Reply:
I think it's good to check that filename is valid, not just empty (e.g. not a folder)